internal/component/prometheus/scrape/scrape.go

```diff
@@ -213,26 +228,61 @@ func (c *Component) Run(ctx context.Context) error {
 		case <-c.reloadTargets:
 			c.mut.RLock()
 			var (
-				targets           = c.args.Targets
-				jobName           = c.opts.ID
-				clusteringEnabled = c.args.Clustering.Enabled
+				targets = c.args.Targets
+				jobName = c.opts.ID
+				args    = c.args
 			)
-			if c.args.JobName != "" {
+			c.mut.RUnlock()
+
+			if args.JobName != "" {
 				jobName = c.args.JobName
 			}
-			c.mut.RUnlock()
 
-			promTargets := c.distTargets(targets, jobName, clusteringEnabled)
+			newTargetGroups, movedTargets := c.distributeTargets(targets, jobName, args)
+
+			// Make sure the targets that moved to another instance are NOT marked as stale. This is specific to how
+			// Prometheus handles marking series as stale: it is the client's responsibility to inject the
+			// staleness markers. In our case, for targets that moved to another instance in the cluster, we hand
+			// over this responsibility to the new owning instance. We must not inject staleness marker here.
+			c.scraper.DisableEndOfRunStalenessMarkers(jobName, movedTargets)
```

Comment:
I haven't looked too closely at all the tests, so this might already be handled, but I want to make sure: does this gracefully handle the case where a target moves to another node, and then moves back to the original node? 

For example, a scenario where the cluster timeline looks like:

```
node-a         # Initial state 
node-a, node-b # node-b joins the cluster 
node-a         # node-b leaves the cluster
```

Reply:
Good point. I don't think I have a full test for this and I suspect that if this happens fast enough, within 5 seconds refresh interval, it may result in some staleness markers being disabled unnecessarily. 

There are two ways to address this:
- Do not update to new targets until the previous targets are properly reloaded and sync'd. I don't think we have a good visibility of this completing now, it's all async.
- We can make sure that we `EnableStalenessMarkers` for all the current targets that we are updating to.
- We can also ignore this edge case - it may be rare and there are already situations where program can crash and staleness will never be injected, so it must be gracefully handled by the TSDB anyway.

Reply:
>  so it must be gracefully handled by the TSDB anyway.

Yes, Prometheus query engines forcibly inject a staleness marker at query time after 4 minutes from the last received sample, so I don't think it's that big of an issue if staleness markers get incorrectly disabled.